### ISSUE TITLE
Llama.cpp now supports distributed inference across multiple machines.

### ISSUE THREAD

**Issue by AncientMystic** [ollama/ollama#4643]
Llama.cpp now supports distribution across multiple devices to boost speeds, this would be a great addition to Ollama 

https://github.com/ggerganov/llama.cpp/tree/master/examples/rpc

https://www.reddit.com/r/LocalLLaMA/comments/1cyzi9e/llamacpp_now_supports_distributed_inference/

**Comment by waywardspooky** [comment]
> I don't think this offers any speedup, yet. It just increases the size of the models you can run. Still useful, though.

Incredibly useful. There's plenty of us that have multiple computers each with their own GPU but for different reasons can't run a machine with multiple GPU's. Adding this support will allow those in the community to run more competent models such as Command-R-Plus, WizardLM-2-8x22B. My workstation has a 3090 ( 24GB ) and a 1080TI ( 11GB ), but I also have an HTPC with a 1080TI ( 11GB ) and another machine with a 1080TI ( 11GB ). With distributed inference I'd have roughly 57 GB of VRAM to work with, which makes a huge different in the quality of responses I can get were I to be able to run larger more capable models.

**Comment by AncientMystic** [comment]
I agree i mean it wouldn't offer a performance boost over the capabilities of the card/cards but for example running a 40gb model on a device with 11gb vram and getting 0.2 t/s vs distributed and having like waywardspooky said 57gb im sure 57gb of vram would do better than trying to run it in ram with cpu so that would be a performance boost in a sense by not suffering a massive drop when running larger models. 

I personally have a SFF pc as a server with a Tesla P4 8GB and a laptop with a 1050Ti 4gb not great really.... but 12GB vram would offer the ability to run better quantisations of 7B models and some 13B models without the drop to 2 t/s or less at least

Most people just don't have a bunch of high vram cards in the same pc and can't afford to so anything at all that can help improve user experiences on lower end devices in anyway is a godsend

**Comment by danividalg** [comment]
Hope this feature can be added to ollama soon :)

**Comment by AncientMystic** [comment]
> Hope this feature can be added to ollama soon :)

Me too, it would really help with using larger models and it would be a lot easier to string together random devices with a few gb of vram then try to afford the top high end GPUs

**Comment by lpb1** [comment]
Hope this feature can be added to ollama soon :)

**Comment by 0x77dev** [comment]
This would be cool in combination with something like Redis to monitor and balance requests. An orchestration layer could ensure models are pulled on specific nodes. Additionally, the ability to specify a custom registry would make Ollama a fully-fledged, enterprise-ready, open-source solution for hosting large language models, but this is out of scope for this issue 🙃

**Comment by saul-jb** [comment]
This would be hugely helpful to get Llama 3.1 405B running on consumer hardware.

**Comment by ahmetegesel** [comment]
I have two MBP 32GB with apple silicon and it would be extremely helpful if ollama implemented this feature. Now that Llama 3.1 is out and tried 70B's capabilities, dying to be able to run its better quants locally.

**Comment by gkpln3** [comment]
I would love to see such a feature in Ollama, self hosting llama3.1:405b can be amazing

**Comment by CesarPetrescu** [comment]
I am looking forward to this. 4 PC's with 4090 yet i cant use multiple on a PC. Would help a lot for llama3.1:70b for me

**Comment by marabgol** [comment]
following, I believe  this is very high demanded! how can we involve and help?

**Comment by Qualzz** [comment]
That's probably the feature i'm waiting the most from ollama.

**Comment by Readon** [comment]
There is also an auto schedule algorithm that [exo](https://github.com/exo-explore/exo) do to mix CPU, GPU resources together.
However, exo is a little bit buggy on intel cpu based machine.

**Comment by kryptech** [comment]
Hello,

I'm on holidays through Sunday during which time I'll be checking my messages rarely and very likely not replying until I'm back at my keyboard.

Regards,
Chris

**Comment by ecyht2** [comment]
I created a PR adding distributed inference, check it out and test if there are any bugs.

**Comment by EvilFreelancer** [comment]
Hi! I've packed `rpc-server` to docker-images with multiple CPU arch supported by Docker.

https://hub.docker.com/r/evilfreelancer/llama.cpp-rpc
https://github.com/EvilFreelancer/docker-llama.cpp-rpc/blob/main/README.en.md

**Comment by hidden1nin** [comment]
Is there a pull request or open development on-going for this?

**Comment by ecyht2** [comment]
> Is there a pull request or open development on-going for this?

Yup, I am working on it. See this PR: #6729.

**Comment by hidden1nin** [comment]
Made a pull request to your pull request, probably didn't do it right but I wanna help!

**Comment by lipere123** [comment]
Hello.
I have a supercomputer and I am heavly using LLama.cpp RPC.
It is working on 6 nodes with one RTX A4000 with 20Go GRAM each.
How can I help to integrate with ollama, please ?
Thanks in advance.
Benjamin.

**Comment by SamuraiBarbi** [comment]
> > Is there a pull request or open development on-going for this?
> 
> Yup, I am working on it. See this PR: #6729.

Any movement on this?

**Comment by mianderson2469** [comment]
I have two Dell poweredge servers each with two nvidia Tesla p40 GPUs.  I would be interested in being able to use the combined resources of these machines even if there is a penalty as long as overall it is a net positive.

**Comment by Kreijstal** [comment]
> I don't think this offers any speedup, yet. It just increases the size of the models you can run. Still useful, though.

how are you running deepseek r1 if it isnt distributed? no quants

**Comment by FayeSpica** [comment]
Waiting for this

**Comment by lipere123** [comment]
![Image](https://github.com/user-attachments/assets/09b3fac2-c393-4b3f-b20a-6fd113970e01) Can be done; Here I am distributed;

**Comment by jli113** [comment]
+1

**Comment by misterjice** [comment]
+1

**Comment by chboishabba** [comment]
+1

**Comment by iAdanos** [comment]
+1

**Comment by ecyht2** [comment]
Sorry for the wait, my PR should be good for testing now.

**Comment by zhangddjs** [comment]
Wow, thank you for the awesome work! And btw may I ask, does the PR support multiple distribute CPUs utilization ? @ecyht2

**Comment by Kreijstal** [comment]
merge when

**Comment by ecyht2** [comment]
> Wow, thank you for the awesome work! And btw may I ask, does the PR support multiple distribute CPUs utilization ? [@ecyht2](https://github.com/ecyht2)

Never tested it but, having 2 different RPC servers running on CPU should work.

**Comment by gkpln3** [comment]
Building on what @ecyht2 started, I'm pretty sure I managed to get it to work, the usage I currently have is:
On slave devices, run:
```
ollama rpc
```

On master device:
```
OLLAMA_RPC_SERVERS="[ip1]:50052,[ip2]:50052" ollama serve
```

I will test it tomorrow, and update here if that works.

**Comment by gkpln3** [comment]
I managed to get an initial version to work. You can find it here: https://github.com/gkpln3/ollama/tree/feat/rpc

**Comment by aquarat** [comment]
@gkpln3 
> I managed to get an initial version to work. You can find it here: https://github.com/gkpln3/ollama/tree/feat/rpc

It unfortunately doesn't build via Docker for me :( but otherwise the changes look good 👍

**Comment by robertgro** [comment]
This is the new link: https://github.com/ggml-org/llama.cpp/tree/master/tools/rpc

**Comment by misterjice** [comment]
So, does this work now for ollama?  olol is not working at all....

**Comment by aquarat** [comment]
I'm using llama.cpp directly.

**Comment by robertschulze** [comment]
> I'm using llama.cpp directly.

But how can you integrate this into the LLM workflow e.g. make it usable via a REST API?

**Comment by aquarat** [comment]
llama.cpp has a server component called llama-server, which exposes the model on an OpenAI compatible endpoint. There is another tool that can dynamically switch models using llama-server as needed (llama-swap), but I don’t use that. I just run one local model currently.

**Comment by overcuriousity** [comment]
Is this feature still being considered for implementation, any status updates?

**Comment by henri9813** [comment]
Hello,

Any news ?

**Comment by nomore1007** [comment]
Been waiting 2 years for this haha

**Comment by regulad** [comment]
Llama.cpp's server binary now supports a router mode.

**Comment by misterjice** [comment]
This is amazing! I've been waiting for this forever. How long before this comes to ollama?  I may have to switch completely to llama CPP at this point.. 

*edit* or not:

### Current Status: **Experimental / "Soft" Implementation**
While Ollama has not yet made "Network Distributed Inference" a primary, one-click feature in the stable CLI (like ollama run), the underlying support has been integrated into the backend.
 * **The Backend is Ready:** Since Ollama uses llama.cpp as its primary engine, the RPC (Remote Procedure Call) capabilities are present in the binary.
 * **Manual Activation:** You can currently trigger this behavior using environment variables. In the latest versions (v0.21.x and above), users have been successful using:
   * **On Worker Machines:** Running ollama rpc (a command added to listen for remote instructions).
   * **On the Main Machine:** Setting OLLAMA_RPC_SERVERS="[IP1]:50052,[IP2]:50052" before starting the service.
### Why isn't it "Official" yet?
The Ollama team is known for prioritizing a "just works" user experience. Distributed inference introduces several headaches that they are still refining:
 1. **Network Latency:** Even on 1Gbps Ethernet, the bottleneck of moving data between machines often makes distributed inference slower than just running a smaller model on a single machine.
 2. **Stability:** If one machine on the network hiccups, the entire inference stream crashes.
 3. **Discovery:** They are working on a way for Ollama instances to "find" each other on a local network automatically (mDNS/Zeroconf) rather than requiring manual IP configuration.
### Summary of the "Change"
| Feature | Status |
|---|---|
| **Multi-GPU (Single Machine)** | Fully Supported (Native) |
| **Distributed (Multi-Machine)** | **Experimental** (Requires CLI environment variables) |
| **Official GUI/Easy-Config** | In Development (Expected late 2026) |


If you have a cluster of machines ready to go, you don't actually have to wait for a "final" release. You can try the OLLAMA_RPC_SERVERS method today, provided your machines are on the same high-speed local network. Just be prepared for a bit of a performance hit compared to local VRAM!

**Comment by nomore1007** [comment]
I tried installing 0.30.0-rc15 on Windows, and it did NOT allow me to run "ollama rpc". 
This is why I have to wait for "Official" support.

**Comment by gkpln3** [comment]
Looping back into this - Apple has announced support for distributed inference across multiple macs using RDMA through Thunderbolt 4 - https://developer.apple.com/videos/play/wwdc2026/233/
Having that implemented natively into Ollama could be amazing.
